### PR TITLE
Credentials 2: Add token class

### PR DESCRIPTION
**Pull request by ianwremmel**

### CODE REVIEW

File: packages/plugin-credentials/src/index.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+import {registerPlugin} from '@ciscospark/spark-core';
+import Credentials from './credentials';
+import config from './config';
+
+registerPlugin(`credentials`, Credentials, {
+  config
+});
+
+export {
+  apiScope,
```
Comment (Josh-Dykstra) [inline packages/plugin-credentials/src/index.js]: I don't see apiScope defined in this file... Might just be that this is a set-up PR and it'll be in the next?

File: packages/plugin-credentials/src/scope.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+ * @returns {string}
+ */
+export function filterScope(toFilter, scope) {
+  if (!scope) {
+    return ``;
+  }
+
+  return scope
+    .split(` `)
+    .filter((item) => item !== `spark:kms`)
```
Comment (Josh-Dykstra) [inline packages/plugin-credentials/src/scope.js]: Should `spark:kms` be passed in as a parameter here? Is there ever another scope we would want to filter

File: packages/plugin-credentials/src/token.js [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+  }
+
+  const ErrorConstructor = grantErrors.select(res.body.error);
+  return Promise.reject(new ErrorConstructor(res._res || res));
+}
+
+const Token = SparkPlugin.extend({
+  namespace: `Credentials`,
+
+  props: {
```
Comment (Josh-Dykstra) [inline packages/plugin-credentials/src/token.js]: I wonder if we can specify any of these as required ampersand props.

File: packages/plugin-credentials/src/token.js [hunk 4]
```diff
@@ -0,0 +1,332 @@
...
+      cache: false,
+      deps: [
+        `expires`
+      ],
+      fn: function isExpired() {
+        return Boolean(this.expires && Date.now() > this.expires);
+      }
+    },
+
+    isUseless: {
```
Comment (Josh-Dykstra) [inline packages/plugin-credentials/src/token.js]: Love it.

File: packages/plugin-credentials/src/token.js [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+      method: `POST`,
+      service: `conversation`,
+      resource: `users/validateAuthToken`,
+      body: {
+        token: this.access_token
+      }
+    })
+      .catch((reason) => {
+        if (reason.message === `\`conversation\` is not a known service`) {
+          const convApi = process.env.CONVERSATION_SERVICE || process.env.CONVERSATION_SERVICE_URL || `https://conv-a.wbx2.com/conversation/api/v1`;
```
Comment (Josh-Dykstra) [inline packages/plugin-credentials/src/token.js]: So if we don't seem to have conversation in the service catalog we try again with a url that's in the environment., and as a last result fallback to this hard-coded url. Should this be something in the defaults file?

Comment (ianwremmel) [inline packages/plugin-credentials/src/token.js]: validate is only for testing. it shouldn't be used in production, so it's fine if we break it in production. similarly, this catch block is also just to handle certain testing scenarios.

Comment (Josh-Dykstra) [inline packages/plugin-credentials/src/token.js]: Got it. Thanks.

File: packages/plugin-credentials/src/scope.js (hunk 2 above)
Comment (ianwremmel) [inline packages/plugin-credentials/src/scope.js]: actually, i passed `toFilter` in for this purpose, just forgot to use it.

File: packages/plugin-credentials/src/token.js (hunk 3 above)
Comment (ianwremmel) [inline packages/plugin-credentials/src/token.js]: not really. a Token is usable if it as as little as just a refresh token or just an access token

**Pull request merged**